[NVPTX] Improve support for {ex2,lg2}.approx

Prince781 opened the pull request:
- Add support for `@llvm.exp2()`:
  - LLVM: `float`        -> PTX: `ex2.approx{.ftz}.f32`
  - LLVM: `half`         -> PTX: `ex2.approx.f16`
  - LLVM: `<2 x half>`   -> PTX: `ex2.approx.f16x2`
  - LLVM: `bfloat`       -> PTX: `ex2.approx.ftz.bf16`
  - LLVM: `<2 x bfloat>` -> PTX: `ex2.approx.ftz.bf16x2`
  - Any operations with non-native vector widths are expanded. On
    targets not supporting f16/bf16, values are promoted to f32.

- Add *CONDITIONAL* support for `@llvm.log2()` [^1]:
  - LLVM: `float` -> PTX: `lg2.approx{.ftz}.f32`
  - Support for f16/bf16 is emulated by promoting values to f32.

[1]: CUDA implements `exp2()` with `ex2.approx` but `log2()` is
implemented differently, so this is off by default. To enable, use the
flag `-nvptx-approx-log2f32`.

Prince781 committed:
[NVPTX] Improve support for {ex2,lg2}.approx

- Add support for `@llvm.exp2()`:
  - LLVM: `float`        -> PTX: `ex2.approx{.ftz}.f32`
  - LLVM: `half`         -> PTX: `ex2.approx.f16`
  - LLVM: `<2 x half>`   -> PTX: `ex2.approx.f16x2`
  - LLVM: `bfloat`       -> PTX: `ex2.approx.ftz.bf16`
  - LLVM: `<2 x bfloat>` -> PTX: `ex2.approx.ftz.bf16x2`
  - Any operations with non-native vector widths are expanded. On
    targets not supporting f16/bf16, values are promoted to f32.

- Add *CONDITIONAL* support for `@llvm.log2()` [^1]:
  - LLVM: `float` -> PTX: `lg2.approx{.ftz}.f32`
  - Support for f16/bf16 is emulated by promoting values to f32.

[1]: CUDA implements `exp2()` with `ex2.approx` but `log2()` is
implemented differently, so this is off by default. To enable, use the
flag `-nvptx-approx-log2f32`.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -518,6 +518,19 @@ multiclass F3_fma_component<string OpcStr, SDNode OpNode> {
+multiclass F1<string OpcStr, SDNode OpNode> {
+   def f32_ftz : NVPTXInst<(outs Float32Regs:$dst), (ins Float32Regs:$a),
+                           !strconcat(OpcStr, ".ftz.f32 \t$dst, $a;"),
+                           [(set Float32Regs:$dst, (OpNode Float32Regs:$a))]>,
```
Please use `f32:` for DAG patterns

Prince781 replied:
Done.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -518,6 +518,19 @@ multiclass F3_fma_component<string OpcStr, SDNode OpNode> {
+// results to zero).
+multiclass F1<string OpcStr, SDNode OpNode> {
+   def f32_ftz : NVPTXInst<(outs Float32Regs:$dst), (ins Float32Regs:$a),
+                           !strconcat(OpcStr, ".ftz.f32 \t$dst, $a;"),
```
Can you please replace the `!strconact` with `#`, it's a bit shorter and more readable?

Prince781 replied:
Done.

AlexMaclean commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,47 @@
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_52 -mattr=+ptx86 | FileCheck --check-prefixes=CHECK %s
+; RUN: %if ptxas-12.6 %{ llc < %s -march=nvptx64 -mcpu=sm_52 -mattr=+ptx86 | %ptxas-verify -arch=sm_52 %}
+source_filename = "fexp2.ll"
+target datalayout = "e-p:64:64:64-p3:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-i128:128:128-f32:32:32-f64:64:64-f128:128:128-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64-a:8:8"
```
Please remove extraneous details such as source_filename and target datalayout from the test-case

AlexMaclean commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,47 @@
+define ptx_kernel void @exp2_test(ptr %a, ptr %res) local_unnamed_addr {
+entry:
+  %in = load float, ptr %a, align 4
+  ; CHECK: ex2.approx.f32 [[D1:%f[0-9]+]], [[S1:%f[0-9]+]]
```
Please use update_llc_test_checks.py to auto-generate these FileCheck patterns.

Prince781 replied:
Thanks!

AlexMaclean commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,47 @@
+target triple = "nvptx64-nvidia-cuda"
+
+; CHECK-LABEL: exp2_test
+define ptx_kernel void @exp2_test(ptr %a, ptr %res) local_unnamed_addr {
```
Instead of kernels can these functions be made device functions which return the result? This way the load and store are not necessary.

Prince781 replied:
Done. Thanks!

AlexMaclean reviewed (commented):
Nice, barring some minor stylistic issues to cleanup this looks good to me. Any chance you could add the (b)f16 variants as well?

durga4github commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,47 @@
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_52 -mattr=+ptx86 | FileCheck --check-prefixes=CHECK %s
+; RUN: %if ptxas-12.6 %{ llc < %s -march=nvptx64 -mcpu=sm_52 -mattr=+ptx86 | %ptxas-verify -arch=sm_52 %}
```
optional: Nothing wrong with using 12.6 version,
Can we use a lesser ptx version since the instruction is available from sm_52 onwards (to get wider coverage)?
(let's say 12.0 or something)

Prince781 replied:
Changed to `ptxas-12.0` and using PTX 1.4.

AlexMaclean reviewed (approved):
Nice, LGTM

durga4github commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,61 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_52 -mattr=+ptx86 | FileCheck --check-prefixes=CHECK %s
+; RUN: %if ptxas-12.0 %{ llc < %s -march=nvptx64 -mcpu=sm_20 -mattr=+ptx14 | %ptxas-verify -arch=sm_20 %}
```
ok, thanks for using ptxas-12.0

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -968,7 +968,8 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
 
-  // No FEXP2, FLOG2.  The PTX ex2 and log2 functions are always approximate.
+  setOperationAction(ISD::FEXP2, MVT::f32, Legal);
+  // No FLOG2. The PTX log2 function is always approximate.
```
`ex2.approx` is also approximate. If not having precise op is a good enough reason not to make log2 legal, why does the same reasoning not apply to exp2?

Prince781 replied:
Perhaps we can add `log2` as well, but that would be for a different commit, no?

Prince781 replied:
Added `log2`.

Artem-B reviewed (commented):
I'm not sure that lowering fexp2 to `ex2.approx` is a good idea.

At the very least it should've been conditional to some sort of fast math flag allowing reduced precision.

Prince781 commented:
> I'm not sure that lowering fexp2 to `ex2.approx` is a good idea.
> 
> At the very least it should've been conditional to some sort of fast math flag allowing reduced precision.

I think it's not a bad idea since there is no non-approximate implementation in PTX, which is something users of NVPTX should know. Making the lowering only work for fast-math would break unoptimized code.

Artem-B commented:
We have explicit flags to enable  approximate reciprocal and sqrt and these instructions should follow a similar pattern.

I agree that enabling them automatically for fast-math may be confusing (though it may be worth checking if we have similar situations on other platforms that could give us some guidelines on how to handle this)

Letting the user enable these instructions explicitly should work.

Letting compiler generate low-precision results will likely break things at runtime (there's a lot of existing code assuming that host/device compilations will produce nearly identical result). I'd prefer things to fail early, in a painfully obvious way if compiler can't do something correctly.

Prince781 commented:
Okay, this feature is now behind the flags `-nvptx-approx-exp2f32` and `-nvptx-approx-log2f32`, which are off by default.

justinfargnoli commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,61 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_20 -mattr=+ptx32 -nvptx-approx-exp2f32 | FileCheck --check-prefixes=CHECK %s
+; RUN: %if ptxas-12.0 %{ llc < %s -march=nvptx64 -mcpu=sm_20 -mattr=+ptx86 -nvptx-approx-exp2f32 | %ptxas-verify -arch=sm_20 %}
```
Why is the PTX version different between these two `llc` invocations?

Prince781 replied:
Fixed. Thanks.

Artem-B commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,61 @@
+; CHECK-EMPTY:
+; CHECK-NEXT:  // %bb.0: // %entry
+; CHECK-NEXT:    ld.param.f32 %f1, [exp2_test_param_0];
+; CHECK-NEXT:    ex2.approx.f32 %f2, %f1;
```
It would be good to see a test that verifies that we do not use those instructions without the flag enabling them.

I guess it would currently still crash LLVM because we can't lower them.

Prince781 replied:
Yeah, I can't think of a good way to have a test for that.

Artem-B replied:
You cound add `; RUN: not llc ...` to make sure it does not compile, and, possibly filecheck for the error message.

Prince781 replied:
Okay, added this.

Prince781 commented:
Updates:

- Added f16 and bf16 variants which promote to f32
- Support is **off** by default. User turns on with either `-nvptx-approx-{log2,exp2}f32` or `-enable-unsafe-fast-math`
- Added expected failure tests when support is not requested.

Artem-B commented on llvm/test/CodeGen/NVPTX/fexp2.ll:
```diff
@@ -0,0 +1,274 @@
+; RUN: %if ptxas-12.0 %{ llc < %s -march=nvptx64 -mcpu=sm_20 -mattr=+ptx32 -nvptx-approx-exp2f32 | %ptxas-verify -arch=sm_20 %}
+target triple = "nvptx64-nvidia-cuda"
+
+; RUN: not --crash llc < %s -march=nvptx64 -mcpu=sm_20 -mattr=+ptx32 2>&1 | FileCheck --check-prefixes=ERR %s
```
Nit: a comment would be useful here

Prince781 replied:
Added comment.

Artem-B commented:
Hmm. Looking at the front-end, I see that we forward `exp2f` to `__nv_exp2f`:
https://github.com/llvm/llvm-project/blob/6f983f88537415952ec528c42f89f1d5b620fe68/clang/lib/Headers/__clang_cuda_math.h#L112

The interesting part is that `__nv_exp2f` in libdevice implements things via `@llvm.nvvm.ex2.approx...`

```
; Function Attrs: alwaysinline nounwind
define float @__nv_exp2f(float %x) #0 {
  %1 = call i32 @__nvvm_reflect(ptr @.str) #6
  %2 = icmp ne i32 %1, 0
  br i1 %2, label %3, label %5

3:                                                ; preds = %0
  %4 = call float @llvm.nvvm.ex2.approx.ftz.f(float %x) #6
  br label %__exp2f.exit

5:                                                ; preds = %0
  %6 = call float @llvm.nvvm.ex2.approx.f(float %x) #6
  br label %__exp2f.exit

__exp2f.exit:                                     ; preds = %3, %5
  %.0 = phi float [ %4, %3 ], [ %6, %5 ]
  ret float %.0
}
```

Considering that CUDA has been living with that implementation all this time, perhaps *that'* is the way we should handle things here, too. In other words using ex2.approx may be OK unconditionally, after all.

Artem-B commented:
@jhuber6 you've dealt with the math library recently. Do you know what are the expectations for the precision of `exp2f` ?

jhuber6 commented:
> @jhuber6 you've dealt with the math library recently. Do you know what are the expectations for the precision of `exp2f`?

The expected precision isn't really documented in the general sense, but OpenCL has some guidelines, or you could look at what GCC currently offers. OpenCL states that `exp2` should be Within 3 ULP https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#relative-error-as-ulps.

jhuber6 commented:
> Hmm. Looking at the front-end, I see that we forward `exp2f` to `__nv_exp2f`:
> 
> https://github.com/llvm/llvm-project/blob/6f983f88537415952ec528c42f89f1d5b620fe68/clang/lib/Headers/__clang_cuda_math.h#L112
> 
> The interesting part is that `__nv_exp2f` in libdevice implements things via `@llvm.nvvm.ex2.approx...`
> 
> ```
> ; Function Attrs: alwaysinline nounwind
> define float @__nv_exp2f(float %x) #0 {
>   %1 = call i32 @__nvvm_reflect(ptr @.str) #6
>   %2 = icmp ne i32 %1, 0
>   br i1 %2, label %3, label %5
> 
> 3:                                                ; preds = %0
>   %4 = call float @llvm.nvvm.ex2.approx.ftz.f(float %x) #6
>   br label %__exp2f.exit
> 
> 5:                                                ; preds = %0
>   %6 = call float @llvm.nvvm.ex2.approx.f(float %x) #6
>   br label %__exp2f.exit
> 
> __exp2f.exit:                                     ; preds = %3, %5
>   %.0 = phi float [ %4, %3 ], [ %6, %5 ]
>   ret float %.0
> }
> ```
> 
> Considering that CUDA has been living with that implementation all this time, perhaps _that'_ is the way we should handle things here, too. In other words using ex2.approx may be OK unconditionally, after all.

Some day I'd really like to remove this little shim layer and just let `exp2` be `exp2` and provide these via the LLVM libm. Right now the main limitation is due to issues w/ linking libcalls via LTO. Because `llvm.exp2` does not link `exp2` it won't be extracted during LTO time. Then this will make it to the backend, get lowered to the `exp2` libcall, and be unresolved. Either that or error on an unimplemented intrinsic. I think we need some logic to state that `llvm.exp2` extracts `exp2` from static libraries, then for certain backends (AMDGPU, NVPTX) before any other optimization we do IR lowering of `llvm.exp2` to `exp2` and put `nobuiltin` on the call site to prevent any further optimizations from turning it back into the LLVM intrinsic.

Artem-B commented:
NVIDIA's docs only [say ](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html?highlight=precision#floating-point-instructions-ex2)"The maximum absolute error is 2-22.5 for fraction in the primary range.".

However, there's a post on NVIDIA's forum, claiming the current implementation error to be 2.38344 ulp:
https://forums.developer.nvidia.com/t/more-accurate-version-of-exp2f-with-no-change-in-performance/243209

Both nvcc and clang currently translate `exp2f` into `ex2.approx`. I think that should be fine. To be on the safe side, we can keep the gating option, but set the default to be enabled.

Prince781 commented:
Update:

- Lowering to `ex2.approx` and `lg2.approx` now **on** default. Use `-nvptx-approx-exp2f32=false` and `-nvptx-approx-log2f32=false` to disable.

Artem-B commented:
Note that I was specifically talking about `exp2`.

`log2` in CUDA gets forwarded to libdevice and does have an elaborate implementation there. It does *not* get lowered to `lg2.approx`. Approximate log2 should remain disabled by default.

Prince781 commented:
Updated.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -94,6 +94,18 @@ static cl::opt<bool> UsePrecSqrtF32(
     cl::desc("NVPTX Specific: 0 use sqrt.approx, 1 use sqrt.rn."),
     cl::init(true));
 
+static cl::opt<bool> UseApproxExp2F32(
```
This could use a comment why it's enabled by default, but log2 isn't.

Something along the lines of:
```
Approximate exp2 is enabled by default to match existing CUDA behavior, which always uses `ex2.approx` instruction. This is not the case for `log2`, so it's disabled by default.
```

Prince781 replied:
Done.

Artem-B reviewed (approved):
LGTM

Prince781 commented:
Updated.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -1193,6 +1206,10 @@ defm FNEG_H: F2_Support_Half<"neg", fneg>;
 
 defm FSQRT : F2<"sqrt.rn", fsqrt>;
 
+defm FEXP2 : F1<"ex2.approx", fexp2>;
```
Looks like we already have these instructions here: 

https://github.com/llvm/llvm-project/blob/fd784726db70a5155594c32ee839b8807fafd87d/llvm/lib/Target/NVPTX/NVPTXIntrinsics.td#L1252-L1268

Instead of creating a duplicate set, can you use patterns to rewrite these nodes to those existing instructins?

Prince781 replied:
Good catch. And I've added support for `ex2.approx.{f16,f16x2,bf16,bf16x2}` variants.

Prince781 commented:
Updated with more improvements. Please see commit message / first comment for more details!

AlexMaclean commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -964,6 +964,10 @@ let TargetPrefix = "nvvm" in {
       DefaultAttrsIntrinsic<[llvm_half_ty], [llvm_half_ty], [IntrNoMem]>;
   def int_nvvm_ex2_approx_f16x2 :
       DefaultAttrsIntrinsic<[llvm_v2f16_ty], [llvm_v2f16_ty], [IntrNoMem]>;
+  def int_nvvm_ex2_approx_bf16 :
+      DefaultAttrsIntrinsic<[llvm_bfloat_ty], [llvm_bfloat_ty], [IntrNoMem]>;
+  def int_nvvm_ex2_approx_bf16x2 :
+      DefaultAttrsIntrinsic<[llvm_v2bf16_ty], [llvm_v2bf16_ty], [IntrNoMem]>;
```
Can we switch nvvm.ex2.approx to be an overloaded intrinsic? Better yet can we use the generic exp2 intrinsic with the `afn` fast-math-flag to represent this operation?

Prince781 replied:
I think for `log2` we should support `afn` to codegen to `lg2.approx`. But for `exp2`, `afn` wouldn't change anything. And the NVVM intrinsics seem to have convention for suffix names for types (`.f` -> `float`, `.f16x2` -> `<2 x half>`, etc) that I can't achieve by using overloading, so I extended them with the same convention.

AlexMaclean replied:
Are these intrinsics different from `llvm.exp2` than? If not, we should probably remove and auto-upgrade them, not add new ones.

Prince781 replied:
They're not different and I think they could be removed, _but_ we have internal code that relies on them.

I can avoid adding the `bf16` variants then so it'll encourage people to switch to using `llvm.exp2` going forward.

Prince781 replied:
Removed the `bf16` intrinsics.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -145,6 +145,7 @@ def noPTXASUnreachableBug : Predicate<"!Subtarget->hasPTXASUnreachableBug()">;
 def doF32FTZ : Predicate<"useF32FTZ()">;
 def doNoF32FTZ : Predicate<"!useF32FTZ()">;
 def doRsqrtOpt : Predicate<"doRsqrtOpt()">;
+def useApproxLog2 : Predicate<"useApproxLog2()">;
```
Is this predicate required? Won't all Log2 nodes be removed during legalization if useApproxLog2 is not enabled?

Prince781 replied:
I think it's not necessary actually.

Prince781 commented:
If I support `afn` as a way to force `lg2.approx` on individual function calls, even if `nvptx-approx-log2f32` is disabled, I run into the problem where SelectionDAG promoting a `flog2` SDNode operation from `f16` to `f32`  will drop the `afn` modifier. Since there is no user-friendly way to request flags being preserved on promotion, I'd have to write my own tablegen pattern matching `f16 = flog2 afn t` -> `f16 = fp_trunc ( f32 = flog2 afn ( fp_extend t ) )`. Is there a better way of going about this?

Prince781 commented:
@AlexMaclean Tried with the `afn` flag on `@llvm.log2()`. This works only if you don't also have non-native operations that are expanded. e.g. `f16 = flog2 afn t0` will be expanded to `f16 = fptrunc (f32 flog2 (f32 fpextend t0))` where SelectionDAG drops `afn` on the new `f32 flog2` node, causing a crash.

It would be nice if SelectionDAG supported something like "preserve `afn`".

Anyway, I think these changes can be merged now.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -1193,6 +1213,8 @@ defm FNEG_H: F2_Support_Half<"neg", fneg>;
 
 defm FSQRT : F2<"sqrt.rn", fsqrt>;
 
+defm FEXP2_H: F2_Support_Half_2<"ex2.approx", fexp2>;
```
Are the f16 and f16x2 variants needed? won't the Patterns in NVPTXIntrinsics.td already handle this?

Prince781 replied:
Thanks. Removed.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -1255,18 +1255,33 @@ def INT_NVVM_EX2_APPROX_F : F_MATH_1<"ex2.approx.f32 \t$dst, $src0;",
+def : Pat<(fexp2 f32:$a),
+          (INT_NVVM_EX2_APPROX_FTZ_F Float32Regs:$a)>, Requires<[doF32FTZ]>;
+def : Pat<(fexp2 f32:$a),
+          (INT_NVVM_EX2_APPROX_F Float32Regs:$a)>, Requires<[doNoF32FTZ]>;
```
Please remove the redundant register class from the output of these patterns. (ie, `(INT_NVVM_EX2_APPROX_F $a)`)

Prince781 replied:
Done.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -1255,18 +1255,33 @@ def INT_NVVM_EX2_APPROX_F : F_MATH_1<"ex2.approx.f32 \t$dst, $src0;",
 def INT_NVVM_LG2_APPROX_D : F_MATH_1<"lg2.approx.f64 \t$dst, $src0;",
   Float64Regs, Float64Regs, int_nvvm_lg2_approx_d>;
 
+def : Pat<(flog2 f32:$a), (INT_NVVM_LG2_APPROX_FTZ_F Float32Regs:$a)>,
```
Remove redundant register classes here too

Prince781 replied:
Right.

Prince781 commented:
Ping

AlexMaclean reviewed (approved):
LGTM

Prince781 commented:
Pinging one of the code owners to merge this.

Prince781 commented:
Thanks @Artem-B!

Prince781 commented:
@Artem-B 

Also, I thought I'd ask here: do you know how I can gain write access? I emailed Chris Lattner but he didn't respond.

jhuber6 commented:
> @Artem-B
> 
> Also, I thought I'd ask here: do you know how I can gain write access? I emailed Chris Lattner but he didn't respond.

https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access though I believe they're in the process of updating that, something like requiring 5 commits and two existing contributors to +1.

Artem-B commented:
@Prince781 It appears that the tests are generating 32-bit PTX and it's no longer supported by recent CUDA versions.

```
[  1] ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
[  2] ; RUN: llc < %s -mcpu=sm_20 -mattr=+ptx32 | FileCheck --check-prefixes=CHECK %s [OK]
llc < third_party/llvm/llvm-project/llvm/test/CodeGen/NVPTX/f32-lg2.ll -mcpu=sm_20 -mattr=+ptx32 | third_party/llvm/llvm-project/llvm/FileCheck --allow-unused-prefixes --check-prefixes=CHECK third_party/llvm/llvm-project/llvm/test/CodeGen/NVPTX/f32-lg2.ll
[  3] ; RUN: %if ptxas %{ llc < %s -mcpu=sm_20 -mattr=+ptx32 | %ptxas-verify %} [FAIL]
 llc < third_party/llvm/llvm-project/llvm/test/CodeGen/NVPTX/f32-lg2.ll -mcpu=sm_20 -mattr=+ptx32 | third_party/gpus/cuda/_virtual_includes/_stage_runtime/third_party/gpus/cuda/bin/ptxas -arch=sm_60 -c -o /dev/null - 
ptxas warning :  64 Bit host architecture (--machine) being used mismatches with .address_size of 32 bits
ptxas fatal   :  32-Bit compilation is no longer supported
Command failed: exit status 255
```

You can reproduce it by running the tests with `LLVM_PTXAS_EXECUTABLE=/path/to/cuda-12.6.0/bin/ptxas`

Prince781 commented:
@jhuber6 Thank you!
